### ISSUE TITLE
Ignore Null values (with conditions) in a row-wise summation for a DataFrame

### ISSUE THREAD

**Issue by lux5** [pola-rs/polars#9642]
### Research

- [X] I have searched the above polars tags on Stack Overflow for similar questions.

- [ ] I have asked my usage related question on [Stack Overflow](https://stackoverflow.com).

### Question about Polars

I would like to calculate the summation for all columns in a row-wise style with such condition:
Ignore the `Null` values if there is at least one not_null values exist in a row, and have `Null` values otherwise. For example,

```py
df = pl.DataFrame({'A': [1, 2, None, None], 'B': [4, None, 6, None]})
print(df)

# shape: (4, 2)
# ┌──────┬──────┐
# │ A    ┆ B    │
# │ ---  ┆ ---  │
# │ i64  ┆ i64  │
# ╞══════╪══════╡
# │ 1    ┆ 4    │
# │ 2    ┆ null │
# │ null ┆ 6    │
# │ null ┆ null │
# └──────┴──────┘
```

What I would like to have:

```py
# shape: (4, 3)
# ┌──────┬──────┬───────┐
# │ A    ┆ B    ┆ Total │
# │ ---  ┆ ---  ┆ ---   │
# │ i64  ┆ i64  ┆ i64   │
# ╞══════╪══════╪═══════╡
# │ 1    ┆ 4    ┆ 5     │
# │ 2    ┆ null ┆ 2     │
# │ null ┆ 6    ┆ 6     │
# │ null ┆ null ┆ null  │
# └──────┴──────┴───────┘
```

In Pandas, you can do such calculation with `min_count=1`:

```py
df = pd.DataFrame({'A': [1, 2, None, None], 'B': [4, None, 6, None]})
df['Total'] = df.sum(axis=1, min_count=1)
print(df)

#      A    B  Total
# 0  1.0  4.0    5.0
# 1  2.0  NaN    2.0
# 2  NaN  6.0    6.0
# 3  NaN  NaN    NaN
```

I checked the `pl.sum` docstring, and couldn't get information on if `Null` values can be ignored. I have tried `pl.fold`, but I still couldn't get what I need. I was wondering what would be the easiest way to achieve this. Many thank :)

**Comment by cmdlineluser** [comment]
Probably a simpler way, but to get started:
```python
df.with_columns(Total = 
   pl.when(pl.any(pl.all().is_not_null()))
     .then(pl.concat_list(pl.all()).list.sum())
)

# shape: (4, 3)
# ┌──────┬──────┬───────┐
# │ A    ┆ B    ┆ Total │
# │ ---  ┆ ---  ┆ ---   │
# │ i64  ┆ i64  ┆ i64   │
# ╞══════╪══════╪═══════╡
# │ 1    ┆ 4    ┆ 5     │
# │ 2    ┆ null ┆ 2     │
# │ null ┆ 6    ┆ 6     │
# │ null ┆ null ┆ null  │
# └──────┴──────┴───────┘
```

**Comment by lux5** [comment]
> ```python
> df.with_columns(Total = 
>    pl.when(pl.any(pl.all().is_not_null()))
>      .then(pl.concat_list(pl.all()).list.sum())
> )
> ```

Thank you so much for the solution. I may need to read on more the row-wise calculations. Most of the time, I am a bit confused with row-wise or column-wise calculations in the Expressions, such as `pl.agg_func()` can be both row-wise and column-wise...

**Comment by cmdlineluser** [comment]
Yes, adding explicit names e.g. `.sum_horizontal()` was suggested in a recent discussion: https://github.com/pola-rs/polars/issues/8646 

I thought I read a recent issue about the difference in null behaviour between the dataframe `.sum()` method but can't seem to find it:
```python
df.select(pl.sum(pl.all()))

# shape: (4, 1)
# ┌──────┐
# │ sum  │
# │ ---  │
# │ i64  │
# ╞══════╡
# │ 5    │
# │ null │
# │ null │
# │ null │
# └──────┘

df.sum(axis=1)

# shape: (4,)
# Series: 'A' [i64]
# [
# 	5
# 	2
# 	6
# 	0
# ]
```

**Comment by lux5** [comment]
> 

By reading the mentioned post, I am getting even more confused. It seems we don't have `pl.row` or any thing like `.sum_horizontal` ready in the current Polars. Also, the behavior of `Null` value is not consistant. To my understanding, it should not be involved in any aggregation calculations, like `sum`, `mean`, `min` etc., unless we direct the aggregation function what to do with the `Null` values. In my other issue (https://github.com/pola-rs/polars/issues/9539) about the `Null` values in `sort_by`, it seems that `Null` values are always considered as the minimum value, which I think not.

**Comment by avimallu** [comment]
>  Also, the behavior of Null value is not consistant.

For `pl.sum`, `null` values are ignored when performing column-wise summation, but propagated when performing row-wise summation. Therefore, you get:

```py
>>> import polars as pl
>>> df = pl.DataFrame({
...     "a": [1, 2, 3, 4],
...     "b": [5, 5, None, 5]})
>>> print(df.select(pl.col("b").sum()))
shape: (1, 1)
┌─────┐
│ b   │
│ --- │
│ i64 │
╞═════╡
│ 15  │
└─────┘
>>> print(df.select(pl.sum("a", "b").alias("both")))
shape: (4, 1)
┌──────┐
│ both │
│ ---  │
│ i64  │
╞══════╡
│ 6    │
│ 7    │
│ null │
│ 9    │
└──────┘
```

> To my understanding, it should not be involved in any aggregation calculations, like sum, mean, min etc., unless we direct the aggregation function what to do with the Null values. 

Agreed that an argument to control behavior would be nice. For now, I hope the explanation in the above code suffices.

**Comment by lux5** [comment]
Thank you so much for the explanation, [avimallu](https://github.com/avimallu). I see now! I think that the solution in the post has successfully solved my issue :)

**Comment by lux5** [comment]
> > Also, the behavior of Null value is not consistant.
> 
> For `pl.sum`, `null` values are ignored when performing column-wise summation, but propagated when performing row-wise summation. Therefore, you get:
> 
> ```python
> >>> import polars as pl
> >>> df = pl.DataFrame({
> ...     "a": [1, 2, 3, 4],
> ...     "b": [5, 5, None, 5]})
> >>> print(df.select(pl.col("b").sum()))
> shape: (1, 1)
> ┌─────┐
> │ b   │
> │ --- │
> │ i64 │
> ╞═════╡
> │ 15  │
> └─────┘
> >>> print(df.select(pl.sum("a", "b").alias("both")))
> shape: (4, 1)
> ┌──────┐
> │ both │
> │ ---  │
> │ i64  │
> ╞══════╡
> │ 6    │
> │ 7    │
> │ null │
> │ 9    │
> └──────┘
> ```
> 
> > To my understanding, it should not be involved in any aggregation calculations, like sum, mean, min etc., unless we direct the aggregation function what to do with the Null values.
> 
> Agreed that an argument to control behavior would be nice. For now, I hope the explanation in the above code suffices.

**Comment by lux5** [comment]
Just found something else confusing me - the above rule seems not applied to row-wise calculation for `pl.min` and `pl.max`: using the same frame in `df`

```python
print(df.select(pl.col("b").min()))
shape: (1, 1)
┌─────┐
│ b   │
│ --- │
│ i64 │
╞═════╡
│ 5   │
└─────┘

print(df.select(pl.min("a", "b").alias("min_row-wise")))
shape: (4, 1)
┌──────────────┐
│ min_row-wise │
│ ---          │
│ i64          │
╞══════════════╡
│ 1            │
│ 2            │
│ 3            │
│ 4            │
└──────────────┘

print(df.select(pl.col("b").max()))
shape: (1, 1)
┌─────┐
│ b   │
│ --- │
│ i64 │
╞═════╡
│ 5   │
└─────┘

print(df.select(pl.max("a", "b").alias("max_row-wise")))
shape: (4, 1)
┌──────────────┐
│ max_row-wise │
│ ---          │
│ i64          │
╞══════════════╡
│ 5            │
│ 5            │
│ 3            │
│ 5            │
└──────────────┘
```

**Comment by stinodego** [comment]
Closing this in favor of #10016

**Comment by stinodego** [comment]
Duplicate of #10016
